### PR TITLE
fix(terminal): stabilize startup, scrolling, and session controls

### LINKED ISSUE #1738: fix(terminal): pass a strict terminal reference to runtime tab rename
## Problem

Renaming a canonical Terminal tab through the runtime socket returns HTTP 500 instead of updating its name.

Observed while preparing a disposable Preview on d4f20f8. The same dispatch code is present on untouched de815be50, so this is independent of the startup-input repair in #1736.

## Cause

`TerminalRuntimeSocketServer.dispatch` handles `RenameTab` by calling `runtime.renameTab(request.input, request.input)`. The input includes `workspaceId`, `tabId`, `name`, and `baseRevision`. `TerminalWorkspaceRuntime.renameTab` immediately parses its first argument with the strict `TerminalRefSchema`, which permits only `workspaceId` and `tabId`. Consequently name/revision are rejected before the rename operation reaches Zellij/storage.

## Scope and acceptance

- Pass only the reference fields as the first runtime argument; retain name/revision as the mutation payload.
- Add a failing test through the real runtime socket and strict runtime reference boundary, not only a mock that accepts extra fields.
- Verify canonical tab rename, optimistic revision conflict, invalid input and owner/project isolation. Preserve terminal identity, command continuity and input lease.
- Check adjacent RPC dispatch paths for the same strict-reference mismatch without bundling unrelated changes.

Existing title search found historical #580 and #1080 for the older session rename flow, but no current issue/PR covering this canonical socket dispatch mismatch. This is a separate item in the full Terminal reliability audit.

### LINKED ISSUE #1745: fix(terminal): synchronize scrollbar with native Zellij history
A terminal can display older native Zellij history after wheel input while its right-hand scrollbar remains near the bottom. The thumb length can also fail to reflect the amount of native history.

The renderer currently combines xterm scrollback and locally clipped canonical rows. Native Zellij mouse reports can scroll a different history buffer without changing xterm `baseY`/`viewportY`, so the displayed content and scrollbar have different sources of truth. The pinned Zellij CLI subscription includes scrollback only on its initial event; later viewport updates do not include native scroll position or total length.

Required behavior:
- Wheel gestures and scrollbar dragging show the same position in the same history.
- Thumb size reflects the real history plus visible content, including new output.
- Both ends remain reachable in narrow/short windows.
- Full-screen applications retain native mouse reports and selection behavior.
- Observer/writer authorization, canonical dimensions, and existing sessions remain unchanged.
- Validate against a real Zellij-backed terminal, not only a standalone xterm fixture.

Investigation is in progress alongside #1736 and the reliability plan #1743. A disposable plugin spike is testing the native pane-contents API (history above and below the viewport); it has not been added to production code. Replacing wheel events with unconditional local xterm scrolling would break native terminal applications and is not an acceptable fix.

### PR DESCRIPTION
**Pull request by xyhStruggler**
## Summary

Opening or reattaching a Terminal could lose startup input, briefly shift text, or show reconnecting. Small windows could expose empty scroll regions or fail to focus from blank space. Deletion could remove and then restore a row, while strict runtime-reference validation rejected rename and pin payloads.

- Buffer startup input in bounded, ordered queues at both gateway and runtime boundaries; retain project admission and writer authorization.
- Decode Zellij snapshot framing and align replay rows before presentation.
- Fit the visible grid to actual content, unify local scrolling, and forward wheel/focus input from blank viewport areas without breaking native mouse reporting or selection.
- Keep sessions visible as **Deleting…** until acknowledgment, reject stale list responses, and lock competing controls. Electron deletion dialogs cannot dismiss early; failed deletion restores controls.
- Separate strict terminal references from rename, pin, input and resize payloads. Preserve current main's Preview collaborator resource-owner admission.

Addresses #1734. Fixes #1738.

## Explicit follow-up scope

**#1745 remains open for a separate PR, as explicitly requested during Human Review.** Native Zellij history can redraw without advancing xterm's local `viewportY`; the scrollbar therefore still has incorrect position/length for that native history. This PR does not claim native history synchronization. Existing local xterm scrolling and clipping tests cannot establish it. A production native-metrics/drag bridge is deferred.

The independent native-attach `runtime_unavailable` observation in #1734 and the complete reliability matrix in Speckit spec/plan #1743 are also separate from these bounded fixes.

## Validation

Current candidate: `e5d409817f449aa8f0ec592ceaf66b7e27be447e`, including main `8b36039a`. Linux reproduction of the failing clipboard case showed the drag starting over another session button after the grid shifted left. The content-clipping stage used `overflow: hidden`, which still permits internal scrolling; it now uses `clip`, leaving deliberate scrolling to the host. A real-renderer regression fails before the change with both internal offsets at 100 and passes afterward with both at zero.

All seven packaged Electron clipboard tests pass on both macOS and Linux/x64. The earlier geometry-settling test fix addressed a separate last-character timing failure; it did not fix the Linux CI failure on its own. Current-head CI (all four unit shards and required Electron E2E), Docker tests and Claude review pass. Greptile gives 5/5 for this exact head.

- 21 Web terminal component tests pass, including pending control lockout, stale-list fencing, and open-menu/editor recovery.
- 43 Electron terminal component tests pass, including Cancel/Escape/backdrop dismissal during deletion and failure recovery.
- Web and Electron TypeScript, React Doctor for the final changed component, and pattern scan pass; pattern scan reports zero violations and five pre-existing warnings.
- Current production Electron builds pass on macOS and Linux/x64. All four browser renderer fixtures and the native Electron fixture pass, along with 73 focused tests and full repository typecheck. The native Electron fixture passed again after current-head Greptile 5/5; the running local Electron also verified blank-area focus and zero internal grid offsets against the current-head Preview.
- Real Preview Pin/Unpin, fresh readback and Rename pass on current head `e5d40981`, using a temporary synthetic tab removed afterward.
- Current full local macOS suite: 15,946 passed, 35 failed, 46 skipped. Thirty failures match the earlier untouched-base failures; the other five fail to bind Unix sockets beneath macOS's long default temporary directory. Re-running those two files unchanged with `TMPDIR=/tmp` passes all 15 tests. This is not a green full local suite. Required exact-head CI unit shards all pass; Linux E2E is tracked separately.

## Human Review and surface evidence

The reviewer tested owner-only [Preview #1736](https://app.matrix-os.com/vm/pr-1736), accepted the delivered behavior, and explicitly authorized merge after Greptile 5/5 and green CI. Subsequent follow-ups close pending-delete control races and prevent hidden internal grid scrolling during focus. No production rollout is included.

| Surface | Evidence / scope |
|---|---|
| Electron Desktop | User-tested Preview; production build; native pixel-wheel, clipping, local rail, selection and blank-area-focus fixture. |
| Web Desktop | Real browser terminal fixture. |
| Web Canvas | Same fixture at 0.75 Canvas zoom. |
| Web Mobile | Real mobile browser viewport fixture. |
| Native Mobile | No native renderer/geometry changes; shared backend contracts covered. Physical-device validation is not claimed. |

The CI E2E job uploads `terminal-soft-resize-evidence` screenshots for the reviewed commit. These synthetic checks do not close #1745.

Preview publication succeeds, but provisioning has repeatedly returned HTTP 500. Scoped platform deployment updates the existing owner Preview without channel promotion, replacement, or user-data deletion. Verified installed/running bundle is `v2026.09.17-pr1736-35277259483-1-e5d4098`, SHA `e5d409817f449aa8f0ec592ceaf66b7e27be447e`; all ten pre-deploy tabs survived. Gateway, shell, sync agent and terminal runtime are healthy. Fresh Pin/Unpin/Rename/readback checks pass; the temporary synthetic tab was removed afterward.

## Invariants

- **Source of truth:** canonical workspace/tab store, live ownership coordinator, and xterm buffer. Queues/scrollbar state remain transient and connection-local.
- **Lock/transaction scope:** one serialized handler per connection; in-flight batches cannot be coalesced into. No schema or database write changes.
- **Acceptable orphan states:** none introduced. Queues, subscriptions, DOM nodes and styles are cleaned up with their owning lifecycle.
- **Auth source of truth:** existing principal, terminal-reference access, project/Chat admission, attachment mode and writer lease. No protocol-reply bypass.

## Existing large-file extraction plan

The control fix stays within existing ShellCard behavior. Before further structural work, extract ShellCard and row controls/styles into TerminalSessionCard.tsx and pure labels into terminal-session-labels.ts; TerminalSidebarItems.tsx will retain group and collapsed-rail composition.

Greptile rated current head `e5d40981` 5/5 and approved it. Both review threads are resolved. All four required CI unit shards, E2E and CI Results pass; Docker tests and Claude review pass. The optional Preview provisioning job remains HTTP 500, with exact-head scoped deployment and live verification completed as documented above.

### CODE REVIEW

File: shell/src/components/terminal/TerminalSidebarItems.tsx [hunk 1]
```diff
@@ -1294,6 +1294,7 @@ function ShellCard({
                 >
                   {displayName}
                 </button>
+                {deleting ? <span role="status" className="shrink-0 text-xs">Deleting…</span> : null}
```
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply shell/src/components/terminal/TerminalSidebarItems.tsx:1300]: [codex] Accept. I will disable all row controls while deletion is pending, including the name/open button, rename, drag handle, and overflow actions, and add a failing-first regression for the pending state. The existing row will remain visible until acknowledgment; this will not change deletion or authorization semantics.

**Comment by xyhStruggler** [comment]
[codex] The native-history finding is valid and remains explicitly deferred to #1745. The reviewer reproduced it during Human Review and has now explicitly requested that this last scrollbar size/position problem be completed in a separate PR, while this PR lands after the review/CI gates.

The existing Zellij mouse path scrolls native history without advancing xterm viewportY; the original xterm rail also derives from the local buffer. This PR consolidates the two local rails and fixes clipped-grid/wheel dead zones, but does not claim to solve that native/local history synchronization gap. The PR body and spec evidence already distinguish those limits. The follow-up needs authoritative native scroll metrics and drag commands; an isolated native API experiment is recorded in #1745. I will preserve that approved scope rather than introduce a new native bridge into this repair.

I am addressing the separate pending-delete control lockout in this PR and will publish a new tested commit. The Greptile 5/5 requirement still applies; this note is the explicit linked follow-up disposition permitted by the repository review policy, not a claim that the scrollbar issue is fixed.

File: shell/src/components/terminal/TerminalSidebarItems.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply shell/src/components/terminal/TerminalSidebarItems.tsx:1300]: [codex] Fixed in 179cc1bb41d3166a64f514166fa84f57961d9849. Open, rename, drag, overflow and all already-open menu items are disabled during deletion; rename-on-blur is guarded too. Failing-first regression plus menu/editor transition and failure-recovery coverage pass (21 component tests), Web TypeScript passes, and pattern scan has zero violations.

**Comment by xyhStruggler** [comment]
[codex] @greptileai please re-review current head 179cc1bb41d3166a64f514166fa84f57961d9849. Pending-delete control lockout is fixed and covered by 21 passing component tests, including open-menu/editor transitions and recovery. Web TypeScript, pattern scan, production Electron build and all four browser renderer fixtures pass. The PR body contains the five-surface evidence matrix and the explicitly user-approved #1745 native-scrollbar follow-up scope. Please evaluate this exact candidate; the earlier 4/5 remains the recorded score until this review completes.

File: desktop/src/renderer/src/features/terminal/TerminalsTab.tsx [hunk 2]
```diff
@@ -399,6 +399,7 @@ export default function TerminalsTab({
         onPin={(shell, pinned) => {
           if (api) void useShellSessions.getState().patchUiState(api, shell.name, { pinned });
         }}
+        deletingName={deleteTarget && isShellBusy(deleteTarget.name) ? deleteTarget.name : null}
```
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply desktop/src/renderer/src/features/terminal/TerminalsTab.tsx:402]: [codex] Accept. I will keep the pending deletion dialog from being dismissed by Cancel, Escape, or backdrop until the request settles, so its row lock cannot disappear early. I will add a failing-first regression covering these dismissal paths and failure recovery without changing backend deletion semantics.

[the remark this answers is not in the document]
Comment (xyhStruggler) [reply desktop/src/renderer/src/features/terminal/TerminalsTab.tsx:402]: [codex] Fixed in a6949d67a82fec8b28d09287d1d885742ab4b003. Dialog onClose now ignores dismissal while the target is busy, and Cancel is disabled. Failing-first regression covers Cancel, Escape, backdrop and failure recovery; all 43 related Desktop tests, Desktop TypeScript, React Doctor (no issues) and pattern scan pass.

**Comment by xyhStruggler** [comment]
[codex] @greptileai please review the latest head a6949d67a82fec8b28d09287d1d885742ab4b003. The latest finding about dismissing an in-flight Electron deletion is fixed: Cancel is disabled and Escape/backdrop dismissal is ignored until the request settles. A failing-first regression plus 43 related Desktop tests, TypeScript, React Doctor and pattern scan pass. The current-head production build and four browser renderer fixtures pass. Please update the confidence score for this exact head; the footer currently still references 179cc1bb. Native scrollbar position/length remains the explicitly approved separate follow-up #1745, as documented in the PR body.

**Comment by xyhStruggler** [comment]
[codex] @greptileai please review head 40ffb0f81cb76235728219c1684540c5dbe87eb9. Production code is identical to the a6949d67 head you rated 5/5. This follow-up only fixes the CI clipboard test's coordinate race: accessible text appears before the 220 ms startup resize debounce and font/grid layout settle. The test now waits for fonts and a stable measured geometry/resize signature before computing drag anchors. The previous behavior reproduced locally; all seven packaged Electron clipboard tests pass after the change, including cross-terminal selection, paste routing and edge auto-scroll. Exact-head CI is restarting. No clipboard behavior or assertion has been relaxed.

**Comment by xyhStruggler** [comment]
[codex] @greptileai please re-review the latest head `e5d409817f449aa8f0ec592ceaf66b7e27be447e`.

Included latest main `8b36039a`. Linux/x64 reproduction of the clipboard failure showed the grid shifted left and the drag start hit a different session button. The clipped internal grid used `overflow: hidden`, which permits browser-driven scrolling; it now uses `overflow: clip`, preserving outer viewport panning. The new real-renderer regression fails before the change (internal x/y offsets both 100) and passes afterward (both 0).

Validation: all 7 packaged Electron clipboard E2Es pass on Linux/x64 and macOS; native Electron grid fixture and all four browser fixtures pass; 73 focused tests, repository typecheck and pattern scan pass. The full local suite is still running. Fresh CI is running for this head; the earlier geometry-settling test change alone did not fix Linux CI.

Surface matrix and Human Review evidence remain in the PR body. Native history scrollbar position/length remains explicitly deferred to #1745. No production rollout. Will wait for current-head 5/5, green CI, and fresh runtime verification before merging.

**Comment by xyhStruggler** [comment]
[codex] Landing verification for `e5d409817f449aa8f0ec592ceaf66b7e27be447e`:

- Greptile 5/5 and approval cover this exact head; zero unresolved review threads.
- CI run 35277259498 passes all four unit shards, required Electron E2E (including the previously failing clipboard case), and CI Results. Docker and Claude review pass.
- Latest main `8b36039a` is included. No subsequent main update at the landing check.
- Post-review native Electron regression passes. The same-head local Electron verifies blank-area focus and zero hidden grid offsets.
- Exact-head Preview installed/running versions match. All ten existing sessions survive; gateway, shell, sync agent and terminal runtime are healthy. Fresh Pin/Unpin/Rename/readback checks pass; the synthetic tab was removed.
- Preview provisioning still returns HTTP 500; its successful published bundle was deployed through the scoped platform endpoint to the existing owner Preview. No production rollout or check override.

The full local macOS suite has 30 unchanged baseline failures plus five default-temp-directory Unix socket path failures; those two socket test files pass all 15 tests with TMPDIR=/tmp. Required Linux CI is green. Remaining native-history scrollbar position/length work stays in #1745.
